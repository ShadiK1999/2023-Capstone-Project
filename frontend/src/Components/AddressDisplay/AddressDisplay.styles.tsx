import { style } from 'typestyle';

export const Container = style({
  $nest: {
    '& td': {
      borderBottom: '1pt solid #000000',
      paddingTop: '75px',
      paddingBottom: '75px',
    },
    '& th': {
      textAlign: 'start',
      borderBottom: '1pt solid #000000',
      paddingBottom: '36px',
      paddingTop: '36px',
    },
  },
  width: '100%',
  borderCollapse: 'collapse',
});

export const VerifiedBadge = style({
  display: 'flex',
  alignItems: 'center',
  gap: '8px',
});

export const IconDiv = style({
  display: 'flex',
  gap: '34px',
});

export const ViewModalContainer = style({
  display: 'flex',
  flexDirection: 'column',
});

export const DayContainer = style({
  paddingLeft: '12px',
});
