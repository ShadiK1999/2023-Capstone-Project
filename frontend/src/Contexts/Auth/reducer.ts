import { produce } from 'immer';
import { AuthState } from './index';

export type AuthAction =
  | {
      type: 'USER_LOGGED_IN';
      payload: string;
    }
  | {
      type: 'USER_LOGGED_OUT';
    };

export const authReducer = produce((draft: AuthState, action: AuthAction) => {
  switch (action.type) {
    case 'USER_LOGGED_IN':
      draft.jwt = action.payload;
      break;
    case 'USER_LOGGED_OUT':
      draft.jwt = null;
      break;
  }
});
