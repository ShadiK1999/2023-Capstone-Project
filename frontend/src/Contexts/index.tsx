export { HomeProvider } from './Home';
export { AuthProvider } from './Auth';
