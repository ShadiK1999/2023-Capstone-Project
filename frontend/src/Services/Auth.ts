import { Result } from './CommonTypes';

export type RegisterRequest = {
  name: string;
  phoneNumber: string;
  password: string;
  email: string;
};

export type LoginRequest = {
  email: string;
  password: string;
};

type LoginBackendResponse = {
  token: string;
  userId: string;
};
type LoginResponse = Result<string>;
type RegisterResponse = Result<undefined>;

export const login = async ({ email, password }: LoginRequest): Promise<LoginResponse> => {
  const body = {
    email,
    password,
  };

  const response = await fetch('api/auth/login', {
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json',
    },
    method: 'POST',
  });

  if (response.ok) {
    const jwt = (await response.json()) as LoginBackendResponse;
    return { value: jwt.token, ok: true };
  }

  return { ok: false, error: 'Wrong email or password' };
};

export const register = async ({
  email,
  password,
  name,
  phoneNumber,
}: RegisterRequest): Promise<RegisterResponse> => {
  const body = {
    email,
    password,
    name,
    phoneNumber,
  };

  const response = await fetch('api/users', {
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json',
    },
    method: 'POST',
  });

  if (response.ok) {
    return { ok: true, value: undefined };
  }

  return {
    ok: false,
    // TODO: probably get an actual error message from the backend
    error: 'You have already created an account with this email, please try logging in',
  };
};
