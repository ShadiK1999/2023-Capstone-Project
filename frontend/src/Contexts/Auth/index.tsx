import { Dispatch, FC, ReactNode, createContext, useContext, useEffect, useReducer } from 'react';
import { AuthAction, authReducer } from './reducer';
import jwt_decode from 'jwt-decode';
export type AuthState = { jwt: string | null };

export type JwtToken = {
  email: string;
  // Id
  sub: string;
  // Timestamp to expire
  exp: number;
  name: string;
};
// Do not use this context, the intended way to use this context is by using the below "use" hooks,
// not by doing `const auth = useContext(AuthContext)`. This context is only exported for mocking purposes
export const AuthContext = createContext<AuthState | undefined>(undefined);

export const AuthDispatchContext = createContext<Dispatch<AuthAction>>(() => void 0);

type ComponentProps = {
  children: ReactNode;
};

const AuthProvider: FC<ComponentProps> = ({ children }) => {
  const [state, dispatch] = useReducer(authReducer, { jwt: localStorage.getItem('jwt') });

  useEffect(() => {
    if (state.jwt) {
      localStorage.setItem('jwt', state.jwt);
    } else {
      localStorage.removeItem('jwt');
    }
  }, [state.jwt]);

  return (
    <AuthContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>{children}</AuthDispatchContext.Provider>
    </AuthContext.Provider>
  );
};

export function useGetDecodedJwt() {
  const context = useContext(AuthContext);

  if (!context || !context.jwt) {
    throw new Error('useDecodedJwt must be used within a Auth Provider and user must be logged in');
  }

  return jwt_decode<JwtToken>(context.jwt);
}

export function useIsAuthenticated(): boolean {
  const context = useContext(AuthContext);

  if (!context) {
    throw new Error('useIsAuthenticated must be used within a Auth Provider');
  }

  return !!context.jwt;
}

/**
 * Function to get the jwt to use for api requests. It is expected to be null,
 * which is when the jwt has expired. The consumer is expected to handle this null
 * value.
 * @returns the jwt
 */
export function useJwt(): string | null {
  const context = useContext(AuthContext);

  if (!context) {
    throw new Error('useJwt must be used within a Auth Provider and user must be logged in');
  }

  if (!context.jwt) return null;

  return context.jwt;
}

export { AuthProvider };
