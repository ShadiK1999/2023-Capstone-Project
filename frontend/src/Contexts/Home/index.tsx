import { FC, ReactNode, useReducer } from 'react';
import { HomeContext, HomeDispatchContext, initialState } from './state';
import { homeReducer } from './reducer';

type ComponentProps = {
  children: ReactNode;
};

const HomeProvider: FC<ComponentProps> = ({ children }) => {
  const [state, dispatch] = useReducer(homeReducer, initialState);

  return (
    <HomeContext.Provider value={state}>
      <HomeDispatchContext.Provider value={dispatch}>{children}</HomeDispatchContext.Provider>
    </HomeContext.Provider>
  );
};

export { HomeProvider };
