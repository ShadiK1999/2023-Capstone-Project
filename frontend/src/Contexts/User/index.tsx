import { Dispatch, FC, ReactNode, createContext, useContext, useReducer } from 'react';
import { UserAction, userReducer } from './reducer';
import { TrustedNeighbour } from '../../Models/TrustedNeighbour';

export type UserState = {
  user: TrustedNeighbour | 'not-loaded';
};

export const UserContext = createContext<UserState | undefined>(undefined);
export const UserDispatchContext = createContext<Dispatch<UserAction>>(() => void 0);

type ComponentProps = {
  children: ReactNode;
};

const UserProvider: FC<ComponentProps> = ({ children }) => {
  const [state, dispatch] = useReducer(userReducer, {
    user: 'not-loaded',
  });

  return (
    <UserContext.Provider value={state}>
      <UserDispatchContext.Provider value={dispatch}>{children}</UserDispatchContext.Provider>
    </UserContext.Provider>
  );
};

export const useUserContext = () => {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error('useUserContext must be used within UserContext');
  }

  return context;
};

export { UserProvider };
