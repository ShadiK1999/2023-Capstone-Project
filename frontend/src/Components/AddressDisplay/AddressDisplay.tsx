import { FC, useCallback, useContext, useState } from 'react';
import { TrustedPoint } from '../../Models/DeliveryPoint';

import * as styles from './AddressDisplay.styles';
import { AddressRegistrationModal, IconButton, Modal } from '..';

import Eye from './Assets/Eye.svg';
import Edit from './Assets/Edit.svg';
import { UserDispatchContext } from '../../Contexts/User';
import { useJwt } from '../../Contexts/Auth';
import { AddressPost, patchAddress } from '../../Services/User';
import { printTime } from '../../Services/LocationsApiClient';

type ComponentProps = {
  trustedPoint: TrustedPoint;
};

function dayOfWeekAsString(dayIndex: number) {
  return ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'][dayIndex];
}

const AddressDisplay: FC<ComponentProps> = ({ trustedPoint }) => {
  const [isEditModalOpen, setIsEditModalOpen] = useState(false);
  const [isViewModalOpen, setIsViewModalOpen] = useState(false);

  const dispatch = useContext(UserDispatchContext);
  const jwt = useJwt();

  const onEditAddress = useCallback(
    async (address: AddressPost) => {
      if (jwt) {
        await patchAddress(address, jwt, trustedPoint.id);

        dispatch({
          type: 'UPDATED_ADDRESS',
          payload: {
            ...address,
            id: trustedPoint.id,
          },
        });
      }
    },
    [dispatch, jwt, trustedPoint.id],
  );
  return (
    <>
      <table className={styles.Container}>
        <thead>
          <tr>
            <th>Name</th>
            <th>Address</th>
            <th>Status</th>
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Address 1</td>
            <td>{trustedPoint.address}</td>
            <td>
              <span className={styles.VerifiedBadge}>
                <svg
                  width="22"
                  height="22"
                  viewBox="0 0 22 22"
                  fill="none"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle cx="11" cy="11" r="11" fill="#1FE86F" />
                </svg>
                <div>Verified</div>
              </span>
            </td>
            <td>
              <div className={styles.IconDiv}>
                <IconButton
                  onClick={() => {
                    setIsViewModalOpen(true);
                  }}
                >
                  <img src={Eye}></img>
                </IconButton>
                {jwt && (
                  <IconButton
                    onClick={() => {
                      setIsEditModalOpen(true);
                    }}
                  >
                    <img src={Edit}></img>
                  </IconButton>
                )}
              </div>
            </td>
          </tr>
        </tbody>
      </table>
      <AddressRegistrationModal
        isOpen={isEditModalOpen}
        onClose={() => {
          setIsEditModalOpen(false);
        }}
        initialAddress={trustedPoint}
        onEnter={onEditAddress}
      />
      <Modal
        title="Address details"
        isOpen={isViewModalOpen}
        onClose={() => {
          setIsViewModalOpen(false);
        }}
      >
        <div className={styles.ViewModalContainer}>
          <div>Address: {trustedPoint.address}</div>
          <br />
          <div>Opening hours:</div>
          <div className={styles.DayContainer}>
            {[0, 1, 2, 3, 4, 5, 6].map((day) => {
              const dayOfWeek = dayOfWeekAsString(day);
              if (!trustedPoint.hours[day]) return <div key={day}>{dayOfWeek}: Closed</div>;

              const openingTime = printTime(trustedPoint.hours[day].openingTime);
              const closingTime = printTime(trustedPoint.hours[day].closingTime);
              return (
                <div key={day}>
                  {dayOfWeek}: {openingTime} - {closingTime}
                </div>
              );
            })}
          </div>
        </div>
      </Modal>
    </>
  );
};

export { AddressDisplay };
