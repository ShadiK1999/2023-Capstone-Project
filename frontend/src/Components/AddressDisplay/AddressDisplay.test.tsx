import { render } from '@testing-library/react';
import { AddressDisplay } from './AddressDisplay';
import { describe, it, expect } from 'vitest';
import { AuthContext } from '../../Contexts/Auth';

describe('Vitest Snapshot testing suite', () => {
  it('Matches DOM Snapshot', () => {
    const tree = render(
      <AuthContext.Provider value={{ jwt: 'test' }}>
        <AddressDisplay
          trustedPoint={{
            id: '1234',
            location: {
              lat: 0,
              lng: 0,
            },
            hours: {},
            address: '1 elizabeth street, melbourne',
          }}
        />
      </AuthContext.Provider>,
    );
    expect(tree).toMatchSnapshot();
  });
});
