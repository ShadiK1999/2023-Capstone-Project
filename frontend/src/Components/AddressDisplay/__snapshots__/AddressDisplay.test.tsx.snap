// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Vitest Snapshot testing suite > Matches DOM Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="fpn0tq5"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Address
            </th>
            <th>
              Status
            </th>
            <th>
              Action
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Address 1
            </td>
            <td>
              1 elizabeth street, melbourne
            </td>
            <td>
              <span
                class="fs9pw6x"
              >
                <svg
                  fill="none"
                  height="22"
                  viewBox="0 0 22 22"
                  width="22"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="11"
                    cy="11"
                    fill="#1FE86F"
                    r="11"
                  />
                </svg>
                <div>
                  Verified
                </div>
              </span>
            </td>
            <td>
              <div
                class="f148jglm"
              >
                <button
                  class="fiwmq2y"
                >
                  <img
                    src="/src/Components/AddressDisplay/Assets/Eye.svg"
                  />
                </button>
                <button
                  class="fiwmq2y"
                >
                  <img
                    src="/src/Components/AddressDisplay/Assets/Edit.svg"
                  />
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="fpn0tq5"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Address
          </th>
          <th>
            Status
          </th>
          <th>
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Address 1
          </td>
          <td>
            1 elizabeth street, melbourne
          </td>
          <td>
            <span
              class="fs9pw6x"
            >
              <svg
                fill="none"
                height="22"
                viewBox="0 0 22 22"
                width="22"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="11"
                  cy="11"
                  fill="#1FE86F"
                  r="11"
                />
              </svg>
              <div>
                Verified
              </div>
            </span>
          </td>
          <td>
            <div
              class="f148jglm"
            >
              <button
                class="fiwmq2y"
              >
                <img
                  src="/src/Components/AddressDisplay/Assets/Eye.svg"
                />
              </button>
              <button
                class="fiwmq2y"
              >
                <img
                  src="/src/Components/AddressDisplay/Assets/Edit.svg"
                />
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
